File: components/storage.py
import os
import shutil
import psutil
from PyQt6.QtWidgets import (
    QLabel, QWidget, QVBoxLayout, QHBoxLayout, QProgressBar,
    QFrame, QSizePolicy
)
from PyQt6.QtGui import QPixmap, QFont
from PyQt6.QtCore import Qt, QTimer


class StorageManager:
    def __init__(self, parent):
        self.parent = parent
        self.drive_widgets = []

        self.init_ui()
        self.setup_timer()

    def init_ui(self):
        self.storage_box = QWidget(self.parent)
        self.storage_box.setGeometry(455, 100, 700, 590)
        self.storage_box.setStyleSheet("""
            background-color: rgba(255, 255, 255, 0.05);
            border: 2px solid rgba(255, 255, 255, 0.15);
            border-radius: 30px;
        """)
        self.storage_box.setVisible(False)

        self.layout = QVBoxLayout(self.storage_box)
        self.layout.setContentsMargins(60, 60, 60, 60)
        self.layout.setSpacing(20)

        storage_label = QLabel("STORAGE", self.storage_box)
        storage_label.setFont(QFont("ElysiaOSNew", 16))
        storage_label.setStyleSheet("color: white; background: transparent; border: none;")
        storage_label.setAlignment(Qt.AlignmentFlag.AlignCenter)

        self.layout.addWidget(storage_label)

        self.drives_container = QVBoxLayout()
        self.drives_container.setSpacing(20)
        self.layout.addLayout(self.drives_container)

    def show(self):
        self.parent.hide_tiles()
        self.parent.show_back_button()
        self.storage_box.setVisible(True)
        self.update_disks()

    def hide(self):
        self.storage_box.setVisible(False)
        self.parent.hide_back_button()
        self.parent.show_tiles()

    def setup_timer(self):
        self.timer = QTimer()
        self.timer.timeout.connect(self.update_disks)
        self.timer.start(1000)  # every 1000ms

    def update_disks(self):
        # Clear previous widgets
        for widget in self.drive_widgets:
            widget.setParent(None)
        self.drive_widgets.clear()

        partitions = psutil.disk_partitions()

        for part in partitions:
            try:
                usage = shutil.disk_usage(part.mountpoint)
                total = usage.total
                free = usage.free
                used = total - free
                percent = int(used / total * 100)
            except Exception:
                continue

            drive_frame = QFrame()
            drive_layout = QHBoxLayout(drive_frame)
            drive_layout.setSpacing(15)
            drive_layout.setContentsMargins(10, 10, 10, 10)

            # Drive icon
            icon_label = QLabel()
            icon_path = os.path.expanduser("~/Elysia/assets/assets/drive.png")
            if os.path.exists(icon_path):
                pixmap = QPixmap(icon_path).scaled(60, 60, Qt.AspectRatioMode.KeepAspectRatio, Qt.TransformationMode.SmoothTransformation)
                icon_label.setPixmap(pixmap)
            else:
                icon_label.setText("Missing")
            icon_label.setFixedSize(50, 50)
            icon_label.setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px")
            drive_layout.addWidget(icon_label)

            # Info section
            info_container = QVBoxLayout()
            info_container.setSpacing(5)

            drive_label = QLabel(f"{part.device} ({part.mountpoint})")
            drive_label.setFont(QFont("ElysiaOSNew", 11))
            drive_label.setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px")

            size_label = QLabel(f"{self.bytes_to_gb(used)} used / {self.bytes_to_gb(total)} total")
            size_label.setFont(QFont("ElysiaOSNew", 10))
            size_label.setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px")

            progress = QProgressBar()
            progress.setMinimum(0)
            progress.setMaximum(100)
            progress.setValue(percent)
            progress.setTextVisible(True)
            progress.setFormat(f"{self.bytes_to_gb(free)} free")
            progress.setFixedHeight(20)
            progress.setSizePolicy(QSizePolicy.Policy.Expanding, QSizePolicy.Policy.Fixed)
            progress.setStyleSheet("""
                QProgressBar {
                    border: 2px solid rgba(255, 255, 255, 0.1);
                    border-radius: 10px;
                    background-color: rgba(255, 255, 255, 0.05);
                    color: white;
                    text-align: center;
                }
                QProgressBar::chunk {
                    background-color: #e5a7c6;
                    border-radius: 10px;
                }
            """)

            info_container.addWidget(drive_label)
            info_container.addWidget(size_label)
            info_container.addWidget(progress)

            drive_layout.addLayout(info_container)
            self.drives_container.addWidget(drive_frame)
            self.drive_widgets.append(drive_frame)

    def bytes_to_gb(self, bytes_val):
        return f"{bytes_val / (1024 ** 3):.1f} GB"

File: components/Storage.cpp
#include "Storage.h"
#include <QDir>
#include <QFont>
#include <QProcess>
#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QFrame>
#include <QLabel>
#include <QProgressBar>
#include <QTimer>
#include <QFileInfo>
#include <QRegularExpression>
#include <QPixmap>
#include <QDebug>

Storage::Storage(QObject *parent) : QObject(parent) {
    QWidget *parentWidget = qobject_cast<QWidget*>(parent);
    if (!parentWidget) return;
    driveWidgets.clear();
    initUI();
    setupTimer();
}

void Storage::initUI() {
    QWidget *parentWidget = qobject_cast<QWidget*>(parent());
    if (!parentWidget) return;
    storageBox = new QWidget(parentWidget);
    storageBox->setGeometry(455, 100, 700, 590);
    storageBox->setStyleSheet(
        "background-color: rgba(255, 255, 255, 0.05);"
        "border: 2px solid rgba(255, 255, 255, 0.15);"
        "border-radius: 30px;"
    );
    storageBox->setVisible(false);
    QVBoxLayout *mainLayout = new QVBoxLayout(storageBox);
    mainLayout->setContentsMargins(60, 60, 60, 60);
    mainLayout->setSpacing(20);
    QLabel *storageLabel = new QLabel("STORAGE", storageBox);
    storageLabel->setFont(QFont("ElysiaOSNew", 16));
    storageLabel->setStyleSheet("color: white; background: transparent; border: none;");
    storageLabel->setAlignment(Qt::AlignCenter);
    mainLayout->addWidget(storageLabel);
    drivesContainer = new QVBoxLayout();
    drivesContainer->setSpacing(20);
    mainLayout->addLayout(drivesContainer);
}

void Storage::show() {
    QWidget *parentWidget = qobject_cast<QWidget*>(parent());
    if (!parentWidget) return;
    if (parentWidget->metaObject()->indexOfMethod("hideTiles()") != -1)
        QMetaObject::invokeMethod(parentWidget, "hideTiles");
    if (parentWidget->metaObject()->indexOfMethod("showBackButton()") != -1)
        QMetaObject::invokeMethod(parentWidget, "showBackButton");
    storageBox->setVisible(true);
    updateDisks();
    if (updateTimer) updateTimer->start();
}

void Storage::hide() {
    storageBox->setVisible(false);
    if (updateTimer) updateTimer->stop();
    QWidget *parentWidget = qobject_cast<QWidget*>(parent());
    if (!parentWidget) return;
    if (parentWidget->metaObject()->indexOfMethod("hideBackButton()") != -1)
        QMetaObject::invokeMethod(parentWidget, "hideBackButton");
    if (parentWidget->metaObject()->indexOfMethod("showTiles()") != -1)
        QMetaObject::invokeMethod(parentWidget, "showTiles");
}

void Storage::setupTimer() {
    updateTimer = new QTimer(this);
    updateTimer->setInterval(1000);
    connect(updateTimer, &QTimer::timeout, this, &Storage::updateDisks);
}

void Storage::updateDisks() {
    // Clear previous widgets
    for (QWidget *widget : driveWidgets) {
        widget->setParent(nullptr);
        widget->deleteLater();
    }
    driveWidgets.clear();

    // Get partition info using df -T -B1
    QProcess process;
    process.start("df", QStringList() << "-T" << "-B1");
    process.waitForFinished();
    QString output = QString::fromUtf8(process.readAllStandardOutput());
    QStringList lines = output.split('\n', Qt::SkipEmptyParts);
    if (lines.size() < 2) return;
    // Header: Filesystem Type 1B-blocks Used Available Use% Mounted on
    QStringList validTypes = {"ext4", "btrfs", "xfs", "vfat", "ntfs", "fuseblk", "exfat"};
    for (int i = 1; i < lines.size(); ++i) {
        QStringList cols = lines[i].split(QRegularExpression("\\s+"), Qt::SkipEmptyParts);
        if (cols.size() < 7) continue;
        QString device = cols[0];
        QString fstype = cols[1];
        QString mountpoint = cols[6];
        if (!validTypes.contains(fstype)) continue;
        bool ok1, ok2, ok3;
        qint64 total = cols[2].toLongLong(&ok1);
        qint64 used = cols[3].toLongLong(&ok2);
        qint64 free = cols[4].toLongLong(&ok3);
        if (!ok1 || !ok2 || !ok3 || total == 0) continue;
        int percent = (int)(used * 100.0 / total);
        // UI
        QFrame *driveFrame = new QFrame();
        QHBoxLayout *driveLayout = new QHBoxLayout(driveFrame);
        driveLayout->setSpacing(15);
        driveLayout->setContentsMargins(10, 10, 10, 10);
        // Icon
        QLabel *iconLabel = new QLabel();
        QString iconPath = QDir::homePath() + "/.config/Elysia/assets/assets/drive.png";
        if (QFileInfo::exists(iconPath)) {
            QPixmap pixmap(iconPath);
            iconLabel->setPixmap(pixmap.scaled(60, 60, Qt::KeepAspectRatio, Qt::SmoothTransformation));
        } else {
            iconLabel->setText("Missing");
        }
        iconLabel->setFixedSize(50, 50);
        iconLabel->setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px");
        driveLayout->addWidget(iconLabel);
        // Info
        QVBoxLayout *infoContainer = new QVBoxLayout();
        infoContainer->setSpacing(5);
        QLabel *driveLabel = new QLabel(QString("%1 (%2)").arg(device, mountpoint));
        driveLabel->setFont(QFont("ElysiaOSNew", 11));
        driveLabel->setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px");
        QLabel *sizeLabel = new QLabel(QString("%1 used / %2 total").arg(bytesToGB(used), bytesToGB(total)));
        sizeLabel->setFont(QFont("ElysiaOSNew", 10));
        sizeLabel->setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px");
        QProgressBar *progress = new QProgressBar();
        progress->setMinimum(0);
        progress->setMaximum(100);
        progress->setValue(percent);
        progress->setTextVisible(true);
        progress->setFormat(QString("%1 free").arg(bytesToGB(free)));
        progress->setFixedHeight(20);
        progress->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Fixed);
        progress->setStyleSheet(
            "QProgressBar {"
            "    border: 2px solid rgba(255, 255, 255, 0.1);"
            "    border-radius: 10px;"
            "    background-color: rgba(255, 255, 255, 0.05);"
            "    color: white;"
            "    text-align: center;"
            "}"
            "QProgressBar::chunk {"
            "    background-color: #e5a7c6;"
            "    border-radius: 10px;"
            "}"
        );
        infoContainer->addWidget(driveLabel);
        infoContainer->addWidget(sizeLabel);
        infoContainer->addWidget(progress);
        driveLayout->addLayout(infoContainer);
        drivesContainer->addWidget(driveFrame);
        driveWidgets.append(driveFrame);
    }
}

QString Storage::bytesToGB(qint64 bytes) {
    return QString::asprintf("%.1f GB", bytes / (1024.0 * 1024 * 1024));
} 